keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1076,265 @@ the remaining requests in that queue start getting faster service.  In
+`width` - it will also be affecting the latency of processing the request.
+In practise, processing a single request isn't really parallelized. So
+if we scale the overall concurrency limit with number of available cores,
+the request should never be able to effectively eat more than N units.
+
+Given that we should try to stay on the safe side and rather overprovision
+the number of units needed by the request, just hardcoding them for some
+constant N for all LIST requests seem like a reasonable starting point.
+
+Given that the estimation is well separated piece of logic, we can decide
+to replace with much more sophisticated logic later (e.g. whether it is
+served from etcd or from cache, whether it is namespaced or not, etc.).
+
+One more important aspect to resolve is what happens if a given priority
+level doesn't have enough concurrency units assigned to it. To be on the
+safe side we should probably implement borrowing across priority levels
+but until this is done, we will simply cap the `width` at the concurrency
+units assigned to a priority level.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  While the queueing mechanism doesn't require any adjustments,
+now the request in order to start processing has to accumulate `<width>`
+units of concurrency.
+
+The important requirement to recast now is fairness. As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification of the current round-robin dispatching of the tokens across
+queues:
+- as soon as we dispatch the first token (concurrency unit) to a queue
+  with LIST request at its front, all further tokens are also assigned
+  to this queue until the LIST request can be started.
+  Such approach allows us to not block too many tokens and avoid deadlocks
+  if we would be dispatching tokens to multiple LIST requests without having
+  enough of them for a given priority level.
+- however, to ensure fairness (especially over longer period of times)
+  that queue accumulates negative budget (equal to the number of additionally
+  consumed tokens).  In the next rounds of dispatching, if the queue would
+  normally get a token, it will just reduce the negative budget without
+  starting any request from that queue.
+
+### Support for WATCH requests
```

Comment:
Comment for myself:
The reason why I think that watch requests are more important than borrowing is because:
- watch is protecting us from overload [reliability issue]
- borrowing is increasing our utilization [efficiency]
I personally think that reliability is more important.

@deads2k - you asked a similar question during the meeting yesterday and I realized I didn't share my view on it.